### PR TITLE
Add `IExceptionInterceptor`

### PR DESCRIPTION
**Pull request by viceroypenguin**
Purpose of this is to be able to customize the exceptions thrown, _a la_ https://github.com/Giorgi/EntityFramework.Exceptions

### CODE REVIEW

File: Source/LinqToDB/Data/DataConnection.Async.cs [hunk 1]
```diff
@@ -384,18 +384,42 @@ public async ValueTask DisposeAsync()
 
 		protected virtual async Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken)
 		{
-			var result = Option<int>.None;
+			if (((IInterceptable<ICommandInterceptor>)this).Interceptor is { } cInterceptor)
```
[the remark this answers is not in the document]
Comment (igor-tkachev) [reply Source/LinqToDB/Data/DataConnection.Async.cs:387]:
```c#
protected virtual async Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken)
{
	var cInterceptor = ((IInterceptable<ICommandInterceptor>)  this).Interceptor;
	var eInterceptor = ((IInterceptable<IExceptionInterceptor>)this).Interceptor;

	if (cInterceptor != null || eInterceptor != null)
	{
		try
		{
			await using (cInterceptor == null ? null : ActivityService.StartAndConfigureAwait(ActivityID.CommandExecuteNonQueryAsync))
			await using (eInterceptor == null ? null : ActivityService.StartAndConfigureAwait(ActivityID.CommandExecuteNonQueryAsync))
			{
				if (cInterceptor != null)
				{
					var result = await cInterceptor.ExecuteNonQueryAsync(new(this), _command!, Option<int>.None, cancellationToken)
						.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);

					if (result.HasValue)
						return result.Value;
				}

				return await _command!.ExecuteNonQueryAsync(cancellationToken)
					.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
			}
		}
		catch (Exception ex)
		{
			if (eInterceptor != null)
				using (ActivityService.Start(ActivityID.ExceptionInterceptorProcessException))
					eInterceptor.ProcessException(new(this), ex);
			throw;
		}
	}

	await using (ActivityService.StartAndConfigureAwait(ActivityID.CommandExecuteNonQueryAsync))
	{
		return await _command!.ExecuteNonQueryAsync(cancellationToken)
			.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
	}
}
```

File: Source/LinqToDB/Interceptors/AggregatedExceptionInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+using LinqToDB.Tools;
+
+namespace LinqToDB.Interceptors
+{
+	sealed class AggregatedExceptionInterceptor : AggregatedInterceptor<IExceptionInterceptor>, IExceptionInterceptor
```
Comment (igor-tkachev) [inline Source/LinqToDB/Interceptors/AggregatedExceptionInterceptor.cs:7]: Do we need a new class for this? I would move this method to `AggregatedInterceptor`.

**Review by igor-tkachev — CHANGES_REQUESTED** [review]

File: Source/LinqToDB/Interceptors/AggregatedExceptionInterceptor.cs (hunk 2 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Interceptors/AggregatedExceptionInterceptor.cs:7]: Simply following the pattern of the other interceptors. If we want to try and change the pattern, I'm open to that, but that should be a fuller discussion.

File: Source/LinqToDB/Data/DataConnection.Async.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (igor-tkachev) [reply Source/LinqToDB/Data/DataConnection.Async.cs:387]: - a developer can return None, which means l2db should handle this in the regular way,
- if I used both interceptors, I wouldn't do the work of ExceptionInterceptor in CommandInterceptor.

[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/LinqToDB/Data/DataConnection.Async.cs]: Addressed

File: Source/LinqToDB/Interceptors/IExceptionInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace LinqToDB.Interceptors
+{
+	public interface IExceptionInterceptor : IInterceptor
+	{
+		/// <summary>
+		///	Event, Triggered when an exception is thrown while executing a database command/query.
```
Comment (MaceWindu) [inline Source/LinqToDB/Interceptors/IExceptionInterceptor.cs:8]:
```suggestion
		///	Event, triggered when an exception is thrown while executing a database command/query.
```

**Review by MaceWindu — APPROVED** [review]

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**